Sources/Protocols/CellItem/CollectionViewCellItem.swift

```diff
@@ -0,0 +1,325 @@
+//
+//  CollectionViewCellItem.swift
+//
+//  Copyright © 2017 Rosberry. All rights reserved.
+//
+
+import UIKit.UICollectionView
+import ObjectiveC.runtime
+
+// MARK: - CollectionViewCellItem
+
+public protocol CollectionViewCellItem: CollectionViewConfigureCellItem,
+                                        CollectionViewReuseCellItem,
+                                        CollectionViewSizeCellItem,
+                                        CollectionViewGeneralCellItem,
+                                        CollectionViewCellItemDataSource,
+                                        CollectionViewSiblingCellItem {
+    
+}
+
+// MARK: - CollectionViewReuseCellItem
+
+public protocol CollectionViewReuseCellItem: AnyObject {
+    var reuseType: ReuseType { get }
+}
+
+// MARK: - CollectionViewSizeCellItem
+
+public protocol CollectionViewSizeCellItem: AnyObject {
+    func size() -> CGSize
+}
+
+// MARK: - CollectionViewConfigureCellItem
+
+public protocol CollectionViewConfigureCellItem: AnyObject {
+    func configure(_ cell: UICollectionViewCell)
+}
+
+// MARK: - CollectionViewSiblingCellItem
+
+public protocol CollectionViewSiblingCellItem: AnyObject {
+    var collectionView: UICollectionView? { get set }
+    var indexPath: IndexPath? { get set }
+    var sectionItem: CollectionViewSectionItem? { get set }
+}
+
+extension CollectionViewSiblingCellItem {
+    public var collectionView: UICollectionView? {
+        get {
+            if let object = objc_getAssociatedObject(self, &AssociatedKeys.collectionView) as? UICollectionView {
+                return object
+            }
+            printContextWarning("We found out that collectionView property for \(self) is nil")
+            return nil
+        }
+        set {
+            objc_setAssociatedObject(self, &AssociatedKeys.collectionView, newValue, .OBJC_ASSOCIATION_ASSIGN)
+        }
+    }
+    
+    public var indexPath: IndexPath? {
+        get {
+            if let object = objc_getAssociatedObject(self, &AssociatedKeys.indexPath) as? IndexPath {
+                return object
+            }
+            printContextWarning("We found out that indexPath property for \(self) is nil")
+            return nil
+        }
+        set {
+            objc_setAssociatedObject(self, &AssociatedKeys.indexPath, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+        }
+    }
+    
+    public weak var sectionItem: CollectionViewSectionItem? {
+        get {
+            if let object = objc_getAssociatedObject(self, &AssociatedKeys.sectionItem) as? CollectionViewSectionItem {
+                return object
+            }
+            printContextWarning("We found out that sectionItem property for \(self) is nil")
```

Comment:
I think it's better to refer to the property via the `#function` keyword, in this case, you don't mind about renaming - all the error messages will be actual.

Reply:
Unfortunately it isn't work. `#function` keyword doesn't give us unique and stable pointer 🤷‍♂️